vtls: add a MesaLink vtls backend

kevinis opened the pull request:
MesaLink is a TLS library written in 100% Rust, a programming
langauge that guarantees memory safety. This PR adds MesaLink as a vtls
backend for curl.

nickzman commented:
Thanks for starting on this. One thing I don't see at first glance... Could you also add MesaLink support to tests/runtests.pl?

kevinis commented:
@nickzman Sure I'll start working on it.

bagder commented:
It could probably be done nicer, but the openssl backend simply checks if there's an IP address there and skips setting the SNI if so:

https://github.com/curl/curl/blob/39cb7130c3d302f7447937cb4f23d73f92561d3b/lib/vtls/openssl.c#L2604-L2613

kevinis commented:
Thanks @bagder. The problem is that MesaLink requires a SNI for its server certificate verifier. So it does not accept an IP address. `localhost` would work, but `127.0.0.1` would not. Let me see what I can do from here.

bagder commented:
> it does not accept an IP address

Yes? That's exactly what goes for the OpenSSL function call too and why I showed you that piece of code...

bagder commented on lib/vtls/mesalink.c:
```diff
@@ -173,22 +173,24 @@ mesalink_connect_step1(struct connectdata *conn, int sockindex)
-              "ERROR: MesaLink does not accept an IP address as a hostname\n");
-      }
+    ) {
+      SSL_set_tlsext_host_name(BACKEND->handle, "localhost");
```
Why this nasty kludge? Surely there's no support in any standard for this behavior? No name == no SNI!

kevinis replied:
I don't like this nasty thing, either! Fine, I'll implement no-SNI support in MesaLink...

kevinis replied:
No name == no SNI. At least this is true for our dependency `webpki`, which we use to verify certificate. Please see the API signature below. It always requires a DNS name as specified in RFC 5280 Section 7.2.

https://github.com/briansmith/webpki/blob/68fcf3dd3b7f90fc51ebde1397da7b6f5d1e4034/src/webpki.rs#L193-L197

bagder replied:
SNI is a TLS *extension*. If you connect to a host without having a known name that we can't pass one for SNI. It is not an uncommon situation. All other TLS libraries can handle this and just don't set any SNI then.

If your library can't handle a non-SNI handshake then I suggest you simply return an error if there's no name to use - which then makes it impossible to connect to "https://192.168.0.1" etc - even if the user disables cert checking. Don't invent a host name, that's plain wrong.

kevinis replied:
I agree it is plain wrong to invent a host name. So the current backend only allows 127.0.0.1 and [::1] and rejects other IP addresses.

https://github.com/curl/curl/blob/8c65cbd575e3994fc44daaa24960c7dc04e92384/lib/vtls/mesalink.c#L190-L194

kevinis replied:
I agree with you. I don't like inventing hostnames either. Anyhow `webpki` always requires a valid DNS name to verify certificates. To make `webpki` happy, I would have to make a hostname somewhere; either in the MesaLink library or in the vtls backend. The `localhost` kludge is a compromise to make both `webpki` and curl's test cases happy...

bagder replied:
Since you can't negotiate TLS without SNI (which is remarkably non-openssl compliant), you must fail the connection. Than yes, you can't run most of the curl test suite. I can't accept that you put that logic into curl for release code.

We probably then need to add some required feature-bit in the TLS-tests that require non-SNI connects to work so that curl+mesalink skips those. A mesalink build could then use `#ifdef CURLDEBUG` and make up the host name accordingly for testing purposes only so that debug builds can run through more tests.

bagder commented:
The Mesalink repo says it is an "OpenSSL compatible" library. Shouldn't it then be possible to use the existing `vtls/openssl.c` code (with possibly some tweaks) against mesalink? I mean, isn't writing a totally separate file sort of counter to the argument that it is compatible with openssl?

kevinis commented:
As of 0.7.0, MesaLink provides OpenSSL-compatible SSL_CTX, SSL, and ERR APIs. However, the OpenSSL vtls backend uses lots of BIO, DSA, EVP APIs which are not yet implemented in MesaLink. And we removed a few OpenSSL APIs that we believe would harm the memory safety of MesaLink. 

Given this situation, we believe the "possibly some tweaks"  would be too much and a dedicated vtls backend is a realistic starting point. 

Similarly, we implemented a Java Secure Socket Extension (like a vtls backend) for MesaLink to replace BoringSSL in Android.

bagder commented:
Thanks for clarifying. I don't think OpenSSL's API is much to mimic and you'll just end up like this: almost compatible. It has never worked for any other TLS library in the long run.

kevinis commented:
The ci results are inconsistent.
https://travis-ci.com/mesalock-linux/curl/builds/84054676
https://travis-ci.org/curl/curl/builds/425453308

Looks like the errors are irrelevant. Should I try rebasing?

bagder commented:
The ones you see in https://travis-ci.com/mesalock-linux/curl/builds/84054676 are segfaults, which seem serious. But I can't understand the discrepancy...

THe https://travis-ci.org/curl/curl/builds/425453308 is the new DoH test failing with libressl-only. I restarted that to see if it reproduces...

bagder commented:
If you're getting ready for merge (seeing as the CI build is green and all now). Here's a few requests of mine:
1. Squash the patches so that you have the minimum number of commits you want merged.
2. "rebuild certs with 2048-bit RSA keys" seems unrelated to mesalink? I propose that change is done as a separate PR to not get it confused/mixed up with mesalink changes.
3. I still want your "localhost-hack" for SNI to be for CURLDEBUG-builds only and not for release curl
4. I really like you getting the travis job in as well. Very nice!

kevinis commented:
OK. I'll start a separate PR for "rebuild certs with 2048-bit RSA keys". Once it is merged, I'll start another PR for the MesaLink vtls backend itself.
